File: typescript/src/main/vsdk/toolkit/render/SoftIdsWrapper.ts
import { RenderOptions } from "../common/RenderOptions";
import { SglContext } from "../render/sgl/SglContext";
import { Scene } from "../scene/Scene";
import { Patch } from "../skin/Patch";
import { SoftIds } from "./SoftIds";

export class SoftIdsWrapper {
  private sgl: SglContext;

  private init(scene: Scene, renderOptions: RenderOptions): void {
    if (scene.camera === null) {
      this.sgl = new SglContext(1, 1);
      return;
    }

    this.sgl = SoftIds.setupSoftFrameBuffer(scene.camera);
    SoftIds.softRenderPatches(scene, renderOptions, this.sgl);
  }

  public constructor(scene: Scene, renderOptions: RenderOptions) {
    this.sgl = new SglContext(1, 1);
    this.init(scene, renderOptions);
  }

  public getSize(width: number[] | null, height: number[] | null): void {
    if (width !== null && width.length > 0) {
      width[0] = this.sgl.width;
    }
    if (height !== null && height.length > 0) {
      height[0] = this.sgl.height;
    }
  }

  public getPatchAtPixel(x: number, y: number): Patch | null {
    const index = (this.sgl.height - 1 - y) * this.sgl.width + x;
    return this.sgl.patchBuffer[index];
  }
}

File: java/src/main/java/vsdk/toolkit/render/SoftIdsWrapper.java
package vsdk.toolkit.render;

import vsdk.toolkit.material.RenderOptions;
import vsdk.toolkit.render.sgl.SglContext;
import vsdk.toolkit.scene.Scene;
import vsdk.toolkit.environment.geometry.elements.Patch;

public class SoftIdsWrapper {
    private SglContext sgl; // Software rendering context, includes frame buffer

    // Also performs the actual ID rendering
    private void init(Scene scene, RenderOptions renderOptions) {
        sgl = SoftIds.setupSoftFrameBuffer(scene.camera);
        SoftIds.softRenderPatches(scene, renderOptions, sgl);
    }

    public SoftIdsWrapper(Scene scene, RenderOptions renderOptions) {
        sgl = null;
        init(scene, renderOptions);
    }

    public void getSize(long[] width, long[] height) {
        if (width != null && width.length > 0) {
            width[0] = sgl.width;
        }
        if (height != null && height.length > 0) {
            height[0] = sgl.height;
        }
    }

    public Patch getPatchAtPixel(int x, int y) {
        int index = (sgl.height - 1 - y) * sgl.width + x;
        return sgl.patchBuffer[index];
    }
}
